Kernel/PartialOrdering.cpp

```diff
@@ -0,0 +1,648 @@
+  switch (v) {
+    case PoComp::UNKNOWN:
+      return PoComp::UNKNOWN;
+    case PoComp::GREATER:
+      return PoComp::LESS;
+    case PoComp::EQUAL:
+      return PoComp::EQUAL;
+    case PoComp::LESS:
+      return PoComp::GREATER;
+    case PoComp::NGEQ:
+      return PoComp::NLEQ;
+    case PoComp::NLEQ:
+      return PoComp::NGEQ;
+    case PoComp::INCOMPARABLE:
+      return PoComp::INCOMPARABLE;
+  }
+  ASSERTION_VIOLATION;
+}
+
+constexpr PoComp weaken(PoComp v) {
+  switch (v) {
+    case PoComp::GREATER:
+      return PoComp::NLEQ;
+    case PoComp::LESS:
+      return PoComp::NGEQ;
+    default:
+      break;
+  }
+  ASSERTION_VIOLATION;
+}
+
+constexpr PoComp strengthen(PoComp v) {
+  switch (v) {
+    case PoComp::NLEQ:
+      return PoComp::GREATER;
+    case PoComp::NGEQ:
+      return PoComp::LESS;
+    default:
+      break;
+  }
+  ASSERTION_VIOLATION;
+}
+
+bool checkCompatibility(PoComp old, PoComp curr, PoComp& res)
+{
+  if (old == PoComp::UNKNOWN) {
+    res = curr;
+    return true;
+  }
+  switch (curr) {
+    case PoComp::GREATER:
+      res = PoComp::GREATER;
+      return old == PoComp::GREATER || old == PoComp::NLEQ;
+    case PoComp::EQUAL:
+      res = PoComp::EQUAL;
+      return old == PoComp::EQUAL;
+    case PoComp::LESS:
+      res = PoComp::LESS;
+      return old == PoComp::LESS || old == PoComp::NGEQ;
+    case PoComp::INCOMPARABLE:
+      res = PoComp::INCOMPARABLE;
+      return old == PoComp::INCOMPARABLE || old == PoComp::NGEQ || old == PoComp::NLEQ;
+    case PoComp::NGEQ:
+      switch (old) {
+        case PoComp::LESS:
+        case PoComp::INCOMPARABLE:
+        case PoComp::NGEQ:
+          res = old;
+          return true;
+        case PoComp::NLEQ:
+          res = PoComp::INCOMPARABLE;
+          return true;
+        default:
+          return false;
+      }
+      break;
+    case PoComp::NLEQ:
+      switch (old) {
+        case PoComp::GREATER:
+        case PoComp::INCOMPARABLE:
+        case PoComp::NLEQ:
+          res = old;
+          return true;
+        case PoComp::NGEQ:
+          res = PoComp::INCOMPARABLE;
+          return true;
+        default:
+          return false;
+      }
+      break;
+    default:
+      break;
+  }
+  ASSERTION_VIOLATION;
+}
+
+string pocompToInfix(PoComp c) {
+  switch (c) {
+    case PoComp::UNKNOWN:
+      return "?";
+    case PoComp::GREATER:
+      return ">";
+    case PoComp::EQUAL:
+      return "=";
+    case PoComp::LESS:
+      return "<";
+    case PoComp::NGEQ:
+      return "≱";
+    case PoComp::NLEQ:
+      return "≰";
+    case PoComp::INCOMPARABLE:
+      return "⋈";
+  }
+  ASSERTION_VIOLATION;
+}
+
+PartialOrdering::PartialOrdering()
+  : _size(0), _array(nullptr) {}
+
+PartialOrdering::PartialOrdering(const PartialOrdering& other)
+  : _size(other._size), _array(nullptr)
+{
+  size_t arrSize = ((_size - 1) * _size / 2);
+  if (arrSize) {
+    void* mem = ALLOC_KNOWN(arrSize*sizeof(PoComp), "Kernel::PartialOrdering");
```

Comment:
If the manual memory management is really necessary (is it? looks like `std::vector` to me), please replace the `ALLOC` macros, `array_new`, `array_delete` etc with just regular `new[]` and `delete[]`. `memcpy` could also be `std::copy` but nbd.